File: ABC/ABC301-400/ABC366/E/E.py
N, D = map(int, input().split())
x = [0] * N
y = [0] * N
for i in range(N):
	x[i], y[i] = map(int, input().split())
x.sort()
y.sort()

M = 2000001
ca = [0] * (D + 1)
j = 0
ns = sum(x) + N * M

for i in range(-M, M + 1):
	while j < N and x[j] <= i:
		j += 1
	ns += 2 * j - N
	if ns <= D:
		ca[ns] += 1

for i in range(D):
	ca[i + 1] += ca[i]

j = 0
ns = sum(y) + N * M
ans = 0

for i in range(-M, M + 1):
	while j < N and y[j] <= i:
		j += 1
	ns += 2 * j - N
	if ns <= D:
		ans += ca[D - ns]

print(ans)

File: ABC/ABC301-400/ABC366/E/E.cpp
#include <bits/stdc++.h>
using namespace std;
using ll = long long;

int main() {
	int N, D;
	cin >> N >> D;
	vector<int> x(N), y(N);
	for (int i = 0; i < N; i++) {
		cin >> x[i] >> y[i];
	}
	sort(x.begin(), x.end());
	sort(y.begin(), y.end());

	const int M = 2000001;
	vector<int> ca(D + 1);
	int j = 0;
	ll ns = accumulate(x.begin(), x.end(), 0LL) + (ll)N * M;
	for (int i = -M; i <= M; i++) {
		while (j < N && x[j] <= i) {
			j++;
		}
		ns += 2 * j - N;
		if (ns <= D) {
			ca[ns]++;
		}
	}
	for (int i = 0; i < D; i++) {
		ca[i + 1] += ca[i];
	}

	j = 0;
	ns = accumulate(y.begin(), y.end(), 0LL) + (ll)N * M;
	ll ans = 0;
	for (int i = -M; i <= M; i++) {
		while (j < N && y[j] <= i) {
			j++;
		}
		ns += 2 * j - N;
		if (ns <= D) {
			ans += ca[D - ns];
		}
	}
	cout << ans << '\n';
}